### PR TITLE
[sc-405] Refactor built in definitions

### PR DESCRIPTION
**Pull request by LunaAmora**

### CODE REVIEW

File: src/term/mod.rs [hunk 1]
```diff
@@ -274,6 +281,11 @@ impl Book {
   pub fn is_def_name_generated(&self, def_id: DefId) -> bool {
     self.def_names.name(&def_id).map_or(false, |Name(name)| name.contains('$'))
   }
+
+  /// Checks if the definition of the given DefId is a built-in
+  pub fn is_builtin(&self, def_id: DefId) -> bool {
+    matches!(self.defs.get(&def_id).unwrap().rules[0].origin, Origin::Builtin)
```
Comment (developedby) [inline src/term/mod.rs:287]: I'd suggest deriving PartialEq for Origin and using a `...origin == Origin::BuiltIn`. I think it's a bit more intuitive to read

File: src/term/transform/definition_pruning.rs [hunk 2]
```diff
@@ -1,85 +1,150 @@
...
-    let def = self.defs.get(&main).unwrap();
-    used.insert(def.def_id);
-    def.rules[0].body.find_used_definitions(&mut used, &self.defs);
+    // Even if we don't prune all the defs, we need check what built-ins are acessible throgh user code
+    if !prune_all {
+      for (def_id, def) in &self.defs {
+        // This needs to be done for each rule in case this pass it's run from has not encoded the pattern match
+        // E.g.: the `flatten_rules` golden test
+        for rule in &def.rules {
+          if !matches!(rule.origin, Origin::Builtin) {
```
Comment (developedby) [inline src/term/transform/definition_pruning.rs:45]: And here

File: src/term/transform/eta_reduction.rs [hunk 3]
```diff
@@ -8,7 +8,7 @@ impl Book {
       def.assert_no_pattern_matching_rules();
       let rule = &mut def.rules[0];
 
-      if reduce_all || rule.generated {
+      if reduce_all || !matches!(rule.origin, Origin::User) {
```
Comment (developedby) [inline src/term/transform/eta_reduction.rs:11]: here

**Review by developedby — CHANGES_REQUESTED** [review]
Very good

**Review by developedby — APPROVED** [review]

**Review by imaqtkatt — APPROVED** [review]

**Pull request merged**